src/mouse/mod.rs

```diff
@@ -0,0 +1,506 @@
+
+        let new_simple_event = match self.buttons.get(button).position {
+            ButtonPosition::Down(mouse_down_info) if self.is_drag(mouse_down_info.position, self.xy) => {
+                Some(MouseEvent::Drag(MouseDragEvent{
+                    mouse_button: button,
+                    start: MouseButtonDown{ time: mouse_down_info.time, position: mouse_down_info.position },
+                    current: MouseButtonDown{ time: SteadyTime::now(), position: self.xy },
+                    button_released: true
+                }))
+            },
+            ButtonPosition::Down(_) => {
+                Some(MouseEvent::Click(MouseClick{
+                    mouse_button: button,
+                    position: self.xy
+                }))
+            },
+            _ => None
+        };
+
+        self.buttons.update(button, |state| {
+            state.position = ButtonPosition::Up;
+            state.was_just_released = true;
+            state.event = new_simple_event;
+        });
+    }
+
+    /// called when the mouse wheel is scrolled
+    pub fn scroll(&mut self, x: f64, y: f64) {
+        self.maybe_scroll_event = Some(MouseEvent::Scroll(Scroll{
+            x: x,
+            y: y
+        }));
+    }
+
+    /// resets the state of the mouse
+    pub fn reset(&mut self) {
+        for button_state in self.buttons.iter_mut() {
+            button_state.reset_pressed_and_released();
+            button_state.event = None;
+        }
+        self.maybe_scroll_event = None;
+        self.scroll.x = 0.0;
+        self.scroll.y = 0.0;
+    }
+
+    /// Returns any mouse events since the last update.
+    pub fn events(&self) -> MouseEventIterator {
+        MouseEventIterator::new(self)
+    }
+
+    fn is_drag(&self, start_position: Point, end_position: Point) -> bool {
+        distance_between(start_position, end_position) >= self.drag_distance_threshold
+    }
+
+    fn get_drag_event_for_move(&self, button: MouseButton,
+                                current_button_state: &ButtonState,
+                                current_xy: Point) -> Option<MouseEvent> {
+        use self::ButtonPosition::Down;
+        use self::events::MouseEvent::Drag;
+
+        match current_button_state.position {
+            Down(info) if self.is_drag(info.position, current_xy) => {
+                Some(Drag(MouseDragEvent{
+                    start: info,
+                    current: MouseButtonDown{
+                        time: SteadyTime::now(),
+                        position: current_xy
+                    },
+                    mouse_button: button,
+                    button_released: false
+                }))
+            },
+            _ => None
+        }
+    }
+
+}
+
+fn distance_between(a: Point, b: Point) -> Scalar {
+    let dx_2 = (a[0] - b[0]).powi(2);
+    let dy_2 = (a[1] - b[1]).powi(2);
+    (dx_2 + dy_2).abs().sqrt()
+}
+
+impl<'a> ::std::iter::Iterator for MouseEventIterator<'a> {
```

Comment:
The `Iterator` trait is in the prelude - no need to use the `::std::iter::` namespace here.